### PR TITLE
Fix lexer line tracking for lone CR continuations

### PR DESCRIPTION
**Pull request by bitwisecook**
## Summary

Fix a critical bug in the lexer's line/column position tracking for backslash-CR (`\<CR>`) continuations. A lone carriage return (old-Mac line ending) was incorrectly treated as a line break in the incremental `line`/`col` counter, while the `\n`-only line-start index (`_line_starts`) did not record it. This caused tokens after a `\<CR>` continuation to report `start` positions one line below their `end` positions — backwards ranges that violated position invariants.

The fix ensures that **only `\n` is treated as a line break for position tracking**. A lone `\<CR>` continuation now advances the column counter instead of the line counter, keeping the incremental counter in sync with the line-start index. CRLF pairs still correctly advance the line (on the `\n` half).

### Changes

**Lexer (`compiler/parsing/lexer.py`)**
- `_parse_brace()`: Fixed CR handling in backslash escapes to advance column for lone CR, line for CRLF
- `_parse_string()`: Same fix for quoted-word backslash escapes
- `_parse_comment()`: Same fix for comment backslash continuations

**Formatter (`tooling/formatter/formatter.py`)**
- Extracted `_format_once()` helper for single formatting pass
- Wrapped `format_tcl()` with stabilisation loop to guarantee idempotency — iterates to a fixed point (up to 8 passes) for malformed input that reconstructs non-idempotently, returning input unchanged if no fixed point exists

**Minifier (`tooling/minifier/minifier.py`)**
- Added `_minify_body_stable()` wrapper with same stabilisation strategy as formatter
- Fixed `_reconstruct_raw()` call to pass `next_tok` for both quoted and unquoted words — a `$var` immediately followed by a word-char fragment needs `${var}` braces regardless of quoting to prevent semantic changes and preserve idempotency

**Incremental reparser (`compiler/parsing/incremental.py`)**
- Added early bail-out when old segmentation involved error recovery (partial commands) — the window is re-segmented with recovery disabled, so it cannot reproduce recovered commands, and a from-scratch pass would re-run recovery

**Syntax tree builder (`compiler/parsing/syntax/build.py`)**
- Removed incorrect folding of backslash-newline tokens in quoted words — the lexer emits `\<newline>` as an `ESC` token (real content) only in quoted words, not as a separator, so it must be kept as a fragment

**Tests**
- Added `TestBackslashCRPositionConsistency` with comprehensive position invariant checks
- Added `TestMinifierBraceCompressionIdempotent` and `TestMinifierIdempotency` for minifier stability
- Added `TestFormatterIdempotency` for formatter stability
- Added `TestIncrementalRecovery` for recovery-aware incremental reparsing
- Extended `test_syntax_tree.py` corpus with CR continuations, astral/combining unicode, and quoted-word backslash-newline content

**Documentation**
- Updated `docs/design/compiler/lexing-segmentation.md` with line-tracking convention
- Updated `docs/design/contracts/formatter-engine.md` with idempotency contract
- Updated `docs/design/compiler/syntax-tree.md` with extended test corpus notes

## Validation

- All new tests pass (position consistency, idempotency, recovery handling, syntax tree equivalence)
- Existing test suite passes
- Fuzz-finding #533 (backslash-CR edge case) now handled correctly

https://claude.ai/code/session_01UHwp5mXCaKoz8m4RP2b5Ar

### CODE REVIEW

File: tooling/formatter/formatter.py [hunk 1]
```diff
@@ -45,3 +37,44 @@ def format_tcl(source: str, config: FormatterConfig | None = None) -> str:
...
+    if config is None:
+        config = FormatterConfig()
+
+    result = _format_once(source, config)
+    # Fast path: already a fixed point (all valid Tcl, the overwhelming case).
+    if _format_once(result, config) == result:
+        return result
+    # Otherwise search for a fixed point a few passes out.
+    current = result
+    for _ in range(_STABILISE_PASSES):
```
Comment (bitwisecook) [inline tooling/formatter/formatter.py:82]: Non-blocking suggestion: this stabilisation loop could track all previously produced outputs, not just the immediately previous one. The current `nxt == current` check already catches a fixed point, but a two-state or longer cycle will run to `_STABILISE_PASSES` and then silently return the original source. Keeping a `seen` set keyed by the output (or by a hash plus an equality guard) would let us detect transform loops early and distinguish them from a simple non-converging growth case. When that happens, emitting a warning to stderr from the CLI-facing path would make it visible that formatting was left unchanged because the passes are fighting each other.

File: tooling/minifier/minifier.py [hunk 2]
```diff
@@ -333,9 +333,37 @@ def minify_tcl(
...
+    re-mangles, which used to grow the output by a delimiter on every pass.
+    Iterate to a fixed point; if none exists within :data:`_STABILISE_PASSES`,
+    the input is not cleanly minifiable, so return it unchanged rather than
+    mangle or unboundedly grow it.
+    """
+    result = _minify_body(source, dialect=dialect)
+    if _minify_body(result, dialect=dialect) == result:
+        return result
+    current = result
+    for _ in range(_STABILISE_PASSES):
```
Comment (bitwisecook) [inline tooling/minifier/minifier.py:370]: Same cycle-detection suggestion here as for `format_tcl`: track previously produced outputs in addition to checking `nxt == current`. That lets `_minify_body_stable` stop on a repeated earlier output instead of waiting for the pass cap, and it gives the CLI a useful place to warn on stderr that minification hit a transform loop and returned the original/stable fallback.

**Review by bitwisecook — COMMENTED** [review]
Reviewed commit `65174a7e8570883fe9b80e9423ee43666a4b61da`.

I agree with the existing inline finding on `_parse_command`: I reproduced backwards token ranges after command substitutions containing a lone `\r` continuation, e.g. `[\\\rx]\nset y 1` makes the following tokens start one line too low. I did not open a duplicate thread for that same issue.

Focused verification run: `env UV_CACHE_DIR=/tmp/uv-cache uv run --extra dev pytest tests/test_lexer.py tests/test_syntax_tree.py tests/test_incremental_reparse.py tests/test_formatter.py tests/test_minifier.py -q` -> `612 passed in 48.50s` (`/tmp/pr537-focused.log`).

File: tooling/formatter/formatter.py (hunk 1 above)
Comment (bitwisecook) [reply tooling/formatter/formatter.py:82]: Done in `727dba7` — added a `seen` set so an oscillation is caught immediately and is distinguishable from never-converging growth (each pass strictly larger, which the cap still bounds); both fall back to the unchanged source, so idempotency is unchanged.

One adaptation from the literal suggestion: I surface the loop case via `logging.warning` rather than a bare stderr write, because the LSP server (`server/features/formatting.py`) calls `format_tcl` too — `logging` reaches stderr through the default last-resort handler *and* the LSP's own log handler. The growth case stays a `debug` line, since an unbalanced delimiter mid-edit is the common, expected case and a per-format warning would spam the LSP log; the warning is reserved for a genuine transform loop (the passes fighting).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tooling/minifier/minifier.py (hunk 2 above)
Comment (bitwisecook) [reply tooling/minifier/minifier.py:370]: Done in `727dba7` — same `seen`-set cycle detection in `_minify_body_stable`: it stops on a repeated earlier output instead of waiting for the cap, and distinguishes a transform loop from never-converging growth. As with `format_tcl`, the loop case is surfaced via `logging.warning` (which reaches stderr via the default handler, and the LSP log) rather than a bare stderr write, and growth stays a quiet `debug` line to avoid spamming the LSP on the common mid-edit unbalanced-delimiter case.

---
_Generated by [Claude Code](https://claude.ai/code)_
